### DISCUSSION TITLE
Deploy hook status

### DISCUSSION THREAD

**Discussion by martinbavio** [vercel/community#1825] [Help] [answered]
Hi, I'm using the Deploy Hook functionality in order to allow an admin to be able to manually trigger builds from the CMS. Whenever I POST to the proper webhook url I receive in response an object like this one:

```
{
    job: {
        id: "WHATEVER_ID",
        state: "PENDING",
        createdAt: 1583356457796
    }
}
```

So everything is great, but now I'd love to be able to let know the admin know whenever that job is completed. Is there an API endpoint that I'm not being able to find where I could query for a job status with the given ID?

Thanks!

**Comment by paulogdm** [comment, marked as answer]
Hello @marbiano ,

First, let me thank you for experimenting with the newly added GitHub discussions. We are glad you took the time to write this thread.
Yes, you can get the deployment status with the following:

- `GET /v11/now/deployments/:id`
- `GET /v11/now/deployments/get?url=[deployment URL or alias]`
**Source**: https://zeit.co/docs/api#endpoints/deployments/get-a-single-deployment
Don't forget to look for the **right scope** as well: https://zeit.co/docs/api#api-basics/authentication/accessing-resources-owned-by-a-team

**Reply by matteosox** [reply]
Reading through this, it looks like this answer is incorrect, and in fact it is not possible to track the state of a deploy based on the id provided in the response. Is this still correct?

**Reply by amyegan** [reply]
I'm not sure about the job id, but you could use the `createdAt` timestamp to get the deployment id with `https://api.vercel.com/v6/deployments?app={project-name}&since={timestamp}` ([list deployments](https://vercel.com/docs/rest-api#endpoints/deployments/list-deployments)). Then, you could get the `readyState` from `https://api.vercel.com/v13/deployments/{deployment-id}` ([get a deployment by ID or URL](https://vercel.com/docs/rest-api#endpoints/deployments/get-a-deployment-by-id-or-url)). Someone shared a more detailed version of this idea in https://github.com/vercel/vercel/discussions/3875#discussioncomment-133714.

I'll ask the larger team about the job id and find out if there's a way to use it to look up a deployment 🤔

**Reply by amyegan** [reply]
Aside from the above suggestion involving the timestamp and API. You can use [Webhooks](https://vercel.com/docs/integrations/webhooks-overview) to get updates on deployment status. Examples: [github.com/vercel/example-integration](https://github.com/vercel/example-integration) and [github.com/samcx/vercel-integration](https://github.com/samcx/vercel-integration)

**Comment by martinbavio** [comment]
Lovely @paulogdm, this is going to make admin's experience way more enjoyable. Thank you!

**Comment by martinbavio** [comment]
@paulogdm I have an extra question for you. Should the job id I get from the response to the original request be the id of a deployment? Because every time I try to use that id I get a `Deployment not found` kind of response.

**Reply by paulogdm** [reply]
Can you paste an id as an example?

**Reply by coetry** [reply]
@marbiano You're correct. The `id` returned from the webhook response is the id of the webhook itself and not the actual deployment. 

Take a look at our docs for more detail on webhook events and the payloads that are sent for each event: https://zeit.co/docs/api#integrations/webhooks/webhook-events

**Reply by martinbavio** [reply]
Hi @coetry, thanks for jumping in. Want to clarify that I'm not using a webhook but a deploy hook, and based on the endpoint I'm being told to use (`/v12/integrations/deploy`) it doesn't seem like the same thing.

@paulogdm This is an example of what I get in response from the deploy hook endpoint:

````
{ 
    job: {
        id: "T04TGXuyu59MfwTAWZs2",
        state: "PENDING",
        createdAt: 1583433954208
    }
}
````

I've been fetching deployments a couple of seconds after and I see the new deployment, but that id is not being shown anywhere, which makes sense since it's a job id and not the deploy id. Is there any way you could include the related deploy id on that deploy hook response?

**Reply by hatchli** [reply]
@marbiano @paulogdm Is there an update to this? How are we supposed to programmatically grab the deploy id when only the job id is in the response?

**Reply by chrisvariety** [reply]
@paulogdm @coetry  do you have an update here?

**Reply by hynra** [reply]
Is there an update to this?

**Reply by aksperiod** [reply]
+1 here.

**Reply by paulogdm** [reply]
We don't have an update yet. However, we are baking something in our API and I will share some news soon.

**Reply by brendanmckenzie** [reply]
I am also interested in this.

My ideal flow would be that I can trigger a build against the deployment and then have the Vercel API callback to my API with status updates for that build.

DatoCMS's "Build triggers" has a "webhookUrl" for providing status updates like this.

**Reply by anthonysapp** [reply]
My approach to solving this was this:
1. send a post request to the deploy hook
2. poll the deployments api (/v5/now/deployments) on an interval with 
- a limit of 1, 
- the project id, 
- and the meta for the branch (in my case `meta-githubCommitRef=staging` or `meta-githubCommitRef=master`
- a "from" key of some time in the future (you can even use the "job.createdAt" property as the "base time" for this:
so all in all something like this (pseudo-code, not tested):

```javascript
let deployHookUrl = `YOUR_VERCEL_DEPLOY_HOOK_URL`;
let vercelApiToken = `YOUR_VERCEL_API_TOKEN`;
let vercelProjectId = `YOUR_VERCEL_PROJECT_ID`;
let branch = `master`;
let deploy = null;

// headers and fetch options for using the vercel api
let headers = new Headers({
  Authorization: `Bearer ${verceApilToken}`,
});

let fetchOptions = {
  method: "get",
  mode: "cors",
  headers,
};

async function getDeploy(time) {
  // get a list of deployments using a time in the future
  // you may have to call this on an interval to get the results you want
  let deploysJson = await fetch(
    `https://api.vercel.com/v5/now/deployments?projectId=${vercelProjectId}&meta-githubCommitRef=${branch}&limit=1&from=${
      time + 600000 // time + 10 mins
    }`,
    fetchOptions
  ).then((res) => res.json());

  deploy = deploysJson.deployments[0];
  console.log(deploy.state); // QUEUED, BUILDING, READY, ERROR...
}

async function triggerAndTrackDeploy() {
  // post request to the deploy url
  let jobJson = await fetch(deployHookUrl, {
    method: "post",
    mode: "cors",
  }).then((res) => res.json());

  // the resulting "job" object returned from the deploy hook
  let job = jobJson.job;
  // get the time the deploy was created
  let jobTime = job.createdAt;
  // branch you're interested in
  // get an update for the deploy on an interval so you can see what state it's in
  setInterval(() => {
    getDeploy(jobTime);
  }, 2000);
}

triggerAndTrackDeploy();
```

This is by no means a full solution, but hopefully it helps get you started.

**Reply by amirhouieh** [reply]
Hi, I am wondering if there are any updates on this?

**Reply by haideralipunjabi** [reply]
Yes, I am also wondering if we can track the status of a deployment triggered using Deploy Hook

**Reply by dyleeo** [reply]
+1

**Reply by alexander-bozung** [reply]
+1

**Reply by zhangyi921** [reply]
+1
